lib/features/push_notification/presentation/controller/fcm_token_controller.dart

```diff
@@ -32,149 +33,140 @@ class FcmTokenController extends FcmBaseController {
+      _getFirebaseRegistrationFromServer(
+        newFcmToken: FcmToken(newToken),
+        deviceClientId: DeviceClientId(FcmUtils.instance.hashTokenToDeviceId(newToken))
+      );
     } else {
-      _getFCMSubscriptionLocalAction();
+      _getStoredFirebaseRegistrationFromCache();
     }
   }
 
-  void _getFcmTokenFromBackend(String deviceId) {
-    if (_getFirebaseSubscriptionInteractor != null) {
-      consumeState(_getFirebaseSubscriptionInteractor!.execute(deviceId));
+  void _getFirebaseRegistrationFromServer({
+    required DeviceClientId deviceClientId,
+    FcmToken? newFcmToken
+  }) {
+    if (_getFirebaseRegistrationByDeviceIdInteractor != null) {
+      consumeState(
+        _getFirebaseRegistrationByDeviceIdInteractor!.execute(
+          deviceClientId: deviceClientId,
+          newFcmToken: newFcmToken
+        )
+      );
+    } else if (newFcmToken != null) {
+      _registerNewFirebaseRegistrationToken(newFcmToken);
     }
   }
 
-  void _storeSubscriptionAction(FCMSubscription fcmSubscription){
-    if (_storeSubscriptionInteractor != null) {
-      consumeState(_storeSubscriptionInteractor!.execute(fcmSubscription));
+  void _getStoredFirebaseRegistrationFromCache() {
+    if (_getStoredFirebaseRegistrationInteractor != null) {
+      consumeState(_getStoredFirebaseRegistrationInteractor!.execute());
     }
   }
 
-  bool _isTokenExpired(FirebaseExpiredTime? expireTime) {
-    log('FcmTokenController::_isTokenExpired():expireTime: $expireTime');
-    if (expireTime != null) {
-      final expireTimeLocal = expireTime.value.value.toLocal();
-      final currentTime = DateTime.now();
-
-      log('FcmTokenController::_isTokenExpired():expireTimeLocal: $expireTimeLocal');
-      log('FcmTokenController::_isTokenExpired():currentTime: $currentTime');
+  void _storeFirebaseRegistrationToCache(FirebaseRegistration firebaseRegistration){
+    if (_storeFirebaseRegistrationInteractor != null) {
+      consumeState(_storeFirebaseRegistrationInteractor!.execute(firebaseRegistration));
+    }
+  }
 
-      return currentTime.isBefore(expireTimeLocal) &&
-        expireTimeLocal.daysBetween(currentTime) <= limitedTimeToExpire;
+  bool _validateFirebaseRegistrationTokenExpired(FirebaseRegistration firebaseRegistration) {
+    log('FcmTokenController::_validateFirebaseRegistrationTokenExpired():firebaseRegistration: $firebaseRegistration');
+    if (firebaseRegistration.expires != null) {
+      final expireTimeLocal = firebaseRegistration.expires!.value.value.toLocal();
+      final currentTime = DateTime.now();
+      log('FcmTokenController::_validateFirebaseRegistrationTokenExpired():expireTimeLocal: $expireTimeLocal | currentTime: $currentTime');
+      final offsetTime = expireTimeLocal.minutesBetween(currentTime);
+      log('FcmTokenController::_validateFirebaseRegistrationTokenExpired():offsetTime: $offsetTime');
+      return currentTime.isBefore(expireTimeLocal) && offsetTime <= limitedTimeToExpire;
```

Comment:
because `difference` also support negative number, so IMO we only to compare with `limitedTimeToExpire` is ok.

```
final offsetTime = currentTime.minutesBetween(expireTimeLocal);
return offsetTime > limitedTimeToExpire;
```

Reply:
If so, we just need to convert it to a positive number. We can use the `abs` method